Issue/6585 onboarding SL banner UI

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #6585 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds: 

- New "Install WC Shipping" banner UI to order detail
- A feature flag that ensured the banner is only displayed for debug builds

### Testing instructions
Open order detail and check the banner is displayed in `debug` build. Banner UI should match this design 9PGX5zb3VKHMYc7qTIv30B-fi-744%3A26803

### Images
<!-- Include before and after images or gifs when appropriate. -->

<img width="300"  src="https://user-images.githubusercontent.com/2663464/170314781-9abc5012-e534-44fc-9852-4768aaeed3dc.png">

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

JorgeMucientes committed:
Add shipping label banner layout

JorgeMucientes committed:
Add install wc shipping banner to order detail layout

JorgeMucientes committed:
Add missing strings for install WC shipping banner

JorgeMucientes committed:
Add minor UI adjustments over banner elements

JorgeMucientes committed:
Add feature flag to show the banner only in debug builds

JorgeMucientes committed:
Replace deprecated use of sumBy with sumOf

JorgeMucientes committed:
Align button text with description text in WC Shipping banner

JorgeMucientes committed:
Remove unused field

JorgeMucientes committed:
Fix unit tests

JorgeMucientes committed:
Merge branch 'trunk' into issue/6585-onboarding-sl-banner-ui

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailFragment.kt:
```diff
@@ -402,7 +422,7 @@ class OrderDetailFragment : BaseFragment(R.layout.fragment_order_detail), OrderP
     private fun showOrderRefunds(refunds: List<Refund>, order: Order) {
         // display the refunds count in the refunds section
-        val refundsCount = refunds.sumBy { refund -> refund.items.sumBy { it.quantity } }
+        val refundsCount = refunds.sumOf { refund -> refund.items.sumOf { it.quantity } }
```
Unrelated, just replaced deprecated usage of `sumBy` with non deprecated equivalent

JorgeMucientes commented on WooCommerce/src/main/res/layout/fragment_order_detail.xml:
```diff
@@ -67,17 +66,26 @@
+                    android:visibility="gone" />
+
+                <!-- Install WC shipping plugin banner -->
+                <com.woocommerce.android.ui.orders.details.views.OrderDetailInstallWcShippingBanner
+                    android:id="@+id/orderDetail_installWcShippingBanner"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginBottom="@dimen/minor_100"
                     android:visibility="gone"
-                    android:contentDescription="@string/refunds"/>
+                    tools:visibility="visible" />
```
Actual changes here. The rest is code reformatting

JorgeMucientes committed:
Add proper style to banner to support dark mode

JorgeMucientes commented:
Pending to add the illustration for dark mode

JorgeMucientes commented on WooCommerce/src/main/res/layout/fragment_order_detail.xml:
```diff
@@ -67,17 +66,27 @@
+
+                <!-- Install WC shipping plugin banner -->
+                <com.woocommerce.android.ui.orders.details.views.OrderDetailInstallWcShippingBanner
+                    android:id="@+id/orderDetail_installWcShippingBanner"
+                    style="@style/Woo.Card.WithoutPadding"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginBottom="@dimen/minor_100"
                     android:visibility="gone"
-                    android:contentDescription="@string/refunds"/>
+                    tools:visibility="visible" />
```
Actual change here. The rest is code reformatting.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailFragment.kt:
```diff
@@ -299,7 +319,7 @@ class OrderDetailFragment : BaseFragment(R.layout.fragment_order_detail), OrderP
             key = CardReaderPaymentDialogFragment.KEY_CARD_PAYMENT_RESULT,
             entryId = R.id.orderDetailFragment
         ) {
-            if (FeatureFlag.CARD_READER.isEnabled()) {
+            if (CARD_READER.isEnabled()) {
```
np, personally I think like `FeatureFlag.CARD_READER.isEnabled()` is more readable than just `CARD_READER.isEnabled()`, the nesting aspect of the first one makes it clear this is a feature flag.

hichamboushaba commented on WooCommerce/src/main/res/layout/order_detail_install_wc_shipping_banner.xml:
```diff
@@ -0,0 +1,52 @@
+            android:focusable="true"
+            android:orientation="vertical">
+
+            <com.google.android.material.textview.MaterialTextView
+                style="@style/TextAppearance.Woo.Subtitle2"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="@dimen/major_100"
+                android:layout_marginTop="@dimen/major_100"
+                android:focusable="false"
+                android:text="@string/shipping_label_install_wc_shipping_banner_title"
+                android:textStyle="bold" />
+
+            <com.google.android.material.textview.MaterialTextView
+                style="@style/TextAppearance.Woo.Body2"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="@dimen/major_100"
+                android:layout_marginTop="@dimen/minor_100"
+                android:focusable="false"
+                android:text="@string/shipping_label_install_wc_shipping_banner_description" />
+
+            <com.google.android.material.button.MaterialButton
+                style="@style/Woo.Button.TextButton"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:focusable="false"
+                android:text="@string/shipping_label_install_wc_shipping_banner_button"
+                android:textAllCaps="false" />
```
Even though there is no exact Android designs yet in Figma, I think this piece need some adjustments:
1. The title part should be using Subtitle1 IMO, as it's more close to the `17px` size we have in Figma.
2. The color of the body part seems to be off.
3. The bottom margin seems to be more than what we need, I think there are two solutions:
     - Either add a bottom margin to the bottom, but make the ImageView's height `match_parent` (this would be a better approach to make the content the controlling part, and not the image's height.
     - Or add `vertically_centered` as gravity for the outer layout, and remove the top margin of the title's view, this would center things nicely too. 

![Screen Shot 2022-05-27 at 09 24 30](https://user-images.githubusercontent.com/1657201/170663741-b6751b91-af3b-4435-965e-d77b61460dc8.jpg)

hichamboushaba reviewed (commented):
Nice work @JorgeMucientes, I left some comments, but none is blocking.

The unit tests are failing though.

hichamboushaba commented:
I reviewed this, then I noticed it still a `draft`, the review was requested, I don't know if the `draft` status is just an oversight.

JorgeMucientes committed:
Fix minor UI glitches banner

JorgeMucientes committed:
Move feature flag out of OrderDetailsViewModel

JorgeMucientes committed:
Fix tests

hichamboushaba commented on WooCommerce/src/main/res/layout/order_detail_install_wc_shipping_banner.xml:
```diff
@@ -0,0 +1,52 @@
+            android:orientation="vertical">
+
+            <com.google.android.material.textview.MaterialTextView
+                style="@style/TextAppearance.Woo.Subtitle2"
```
A point that I missed in my initial review, I think it's better to set this as `textAppearance` and not style, and also using passing the theme attribute is better than hardcoding the style:

```suggestion
                android:textAppearance="?attr/textAppearanceSubtitle2"
```

JorgeMucientes committed:
Reduce margin top

JorgeMucientes committed:
Merge branch 'trunk' into issue/6585-onboarding-sl-banner-ui

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt

JorgeMucientes commented:
Thanks for checking this @hichamboushaba ! 

> I reviewed this, then I noticed it still a draft, the review was requested

Yes, sorry about this. I initially set the PR ready for review but then move it back to draft as there were a couple things pending (dark mode asset, fixing tests and merge conflicts). In any case your feedback is useful even if this was in draft. 

> personally I think like FeatureFlag.CARD_READER.isEnabled() is more readable 

I agree with this. In fact, that is how I had added the feature flag inside `OrderDetailViewModel`(which now I moved back to the fragment to fix unit test problems). But the habit of hitting `alt + enter` to add static imports tricked me there and then I missed reverting those changes. 

> The title part should be using Subtitle1 IMO, as it's more close to the 17px size we have in Figma.
Good catch. Fixed it. 

> The color of the body part seems to be off.

Yeah, I'm aware the color is off and I had some doubts on whether to override the color. But finally decided to leave it as is for the following reasons: 
- We are working using iOS Figma designs so there's not always a perfect match between text styles between platforms. 
- I prefer to use an already existing text style that has a slightly different color, rather than creating a new one or adding small overrides here and there to match an iOS design. 
- The current color is consistent with the rest of the texts with similar "relevance" in the same screen. 

I'm still open to updating it, but just wanted to provide some context on why I kept the text color as is for `Body2`

> The bottom margin seems to be more than what we need

Well spotted. This is trickier than it looks. The image height is "longer" than the 3 texts together and their margins. And the image must match the top/bottom of the banner. With these restrictions, the only way to have a more or less centered text is to add `layout_gravity="center_vertical"` as suggested in your second option but keeping some `marginTop` for the title. Bear in mind that the 3 text is actually a button, and this affects how centered gravity is calculated. The final result is this: 

![Screenshot 2022-05-27 at 13 50 12](https://user-images.githubusercontent.com/2663464/170693843-c85de0fe-1ad8-4d58-8b5f-092222c9ace7.png)

JorgeMucientes committed:
Use theme attr instead of style

JorgeMucientes commented:
PS: I'll keep this in draft until we have the image for dark mode available.

hichamboushaba commented:
Thanks for the clarifications @JorgeMucientes 

> I initially set the PR ready for review but then move it back to draft as there were a couple things pending (dark mode asset, fixing tests and merge conflicts).

I tested the PR in draft, and it looked nice, so I assumed it was the same asset for both themes 😄.

> The image height is "longer" than the 3 texts together and their margins. And the image must match the top/bottom of the banner. With these restrictions, the only way to have a more or less centered text is to add layout_gravity="center_vertical" as suggested in your second option but keeping some marginTop for the title

Yes, I understand the cause of the issue, and that's why my first suggestion was to make the image `match_parent`, this way it won't impact the layout's height, and instead it will be resized following the content's height, I prefer this solution as it would work nicely even if when using large fonts in the phone.

JorgeMucientes commented:
> that's why my first suggestion was to make the image match_parent, this way it won't impact the layout's height, and instead, it will be resized following the content's height

The thing is that I'm using `layout_weight` to distribute horizontal space in a way that the image occupies only the area required by its drawable and the remaining free space is filled with the text content. The moment I use weight for this, then setting the image to `match_parent` won't have any impact. Removing the `layout_weight` approach would force to distribute the horizontal space in a fixed manner, and I'd rather avoid that. 

And still, if I removed the use of `layout_weight` and distribute horizontal space in a fixed way. Setting the image height to `match_parent` won't remove the extra bottom margin. It'd only solve the problem when the text content is higher than the image which is not the case unless we increase the font size. But if the image is higher than the text content, there'd be extra space at the bottom anyway.

I think is a great point to think about large fonts. I always forget about checking these sorts of things. With the actual implementation scaling the font to the maximum is far from nice aesthetically but is perfectly usable. (This could also happen for other locales with longer texts though, which is actually more concerning)

![Screenshot 2022-05-27 at 17 22 50](https://user-images.githubusercontent.com/2663464/170729877-e14f3f42-309d-4935-9525-af7fb15a9910.png)

I've been thinking about alternative solutions to the current implementation but all have tradeoffs. 

Let me know if I'm missing something (which could perfectly be the case 😅 )  or if you have another alternative that could work for all cases.

hichamboushaba commented:
> The thing is that I'm using layout_weight to distribute horizontal space in a way that the image occupies only the area required by its drawable and the remaining free space is filled with the text content. The moment I use weight for this, then setting the image to match_parent won't have any impact. Removing the layout_weight approach would force to distribute the horizontal space in a fixed manner, and I'd rather avoid that.

But you are setting the weight on the `width`, and here I'm suggesting setting `match_parent` on the `height`, they shouldn't impact each other.

To make what I mean clearer, this is it in a patch:
```patch
Index: WooCommerce/src/main/res/layout/order_detail_install_wc_shipping_banner.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/res/layout/order_detail_install_wc_shipping_banner.xml b/WooCommerce/src/main/res/layout/order_detail_install_wc_shipping_banner.xml
--- a/WooCommerce/src/main/res/layout/order_detail_install_wc_shipping_banner.xml	(revision e507ee7106a83735aebed1273fc102fc90b77ad4)
+++ b/WooCommerce/src/main/res/layout/order_detail_install_wc_shipping_banner.xml	(date 1653665718124)
@@ -36,6 +36,7 @@
                 style="@style/Woo.Button.TextButton"
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
+                android:layout_marginBottom="@dimen/minor_100"
                 android:focusable="false"
                 android:text="@string/shipping_label_install_wc_shipping_banner_button"
                 android:textAllCaps="false" />
@@ -44,8 +45,9 @@
 
         <ImageView
             android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
+            android:layout_height="match_parent"
             android:importantForAccessibility="no"
+            android:scaleType="fitCenter"
             android:src="@drawable/img_shipping_labels_banner" />
 
     </LinearLayout>
```

\* The `marginBottom` of the bottom to be confirmed if it should be `minor_100` or less, to my eyes, it looks good, but to be confirmed.

---------

The above patch would give the following result:
<img width="400" alt="Screen Shot 2022-05-27 at 16 32 21" src="https://user-images.githubusercontent.com/1657201/170732336-817095a5-9c4e-4ab3-8534-16ec6655affc.png">

And if the text was to get longer, the result would be:
<img width="400" alt="Screen Shot 2022-05-27 at 16 35 10" src="https://user-images.githubusercontent.com/1657201/170732400-60ec0991-d0ab-4283-816a-bb0cf9bc9866.png">

Or alternatively with `fitEnd` (the above snippet applies `fitCenter`) if it matches the design better (to be checked):
<img width="400" alt="Screen Shot 2022-05-27 at 16 36 39" src="https://user-images.githubusercontent.com/1657201/170732533-cef92cd0-f0dc-4d98-92c9-b4cada14b9c6.png">

I suggested above that the image would resize with the content's height, but I just remembered that if the image gets resized, it would take from the text's width, so I applied a `fit` scaling instead, but the rest still applies.

JorgeMucientes committed:
Fix minor UI glitches

JorgeMucientes commented:
Thank you for the explanations @hichamboushaba 🙏🏼  All clear now!

> But you are setting the weight on the width, and here I'm suggesting setting match_parent on the height, they shouldn't impact each other.

Yeah I was referring to `match_parent` for height too 🙂. What I tried to explain was that using `layout_weight` to set the width of the image will prevent the image from resizing even when setting `match_parent` for the height. And I kind of understood that your proposed solution was aiming to achieve that: to resize the image so that it always matched the top and bottom borders of the banner if the text content grew in height. And that had me torn out 😅 , as I could not see how that was feasible with the proposed changes. 

Great suggestion about using `fitEnd` I think it works better for long texts. 

> The marginBottom of the bottom to be confirmed if it should be minor_100 or less

I think it might be less, but the differences are very subtle and the final result is totally fine imo. But I'll leave that judgement to the eye of  @joe-keenan. Let us know what you think.

Final results: 

![Screenshot 2022-05-27 at 18 27 32](https://user-images.githubusercontent.com/2663464/170740528-f75a648e-c39b-4b0f-88c0-65ac52076e4f.png)

With long text the image will be anchored to the bottom

![Screenshot 2022-05-27 at 18 29 43](https://user-images.githubusercontent.com/2663464/170740724-9bfb5f87-4d20-4e15-8a0e-092cc4228cff.png)

hichamboushaba commented:
> And I kind of understood that your proposed solution was aiming to achieve that: to resize the image so that it always matched the top and bottom borders of the banner if the text content grew in height. And that had me torn out 😅 , as I could not see how that was feasible with the proposed changes.

Sorry for the confusion, the goal of my suggestion was to make the layout to depend on the content's height, and not the illustration, but I see how mentioning "resizing" would cause the confusion 😄, to be fair, the image would still resize when the content's height is less that the image's default height 😅.
Thanks for the discussion @JorgeMucientes.

joe-keenan commented:
Hey @hichamboushaba and @JorgeMucientes!
I think you’ve got the text resizing and positioning pretty well handled. From the examples in your comments above, vertically centring the image looks the most natural to me.

I’ve realised that I need to supply you with an unclipped version of the image, though! I always forget that Figma doesn’t make exporting a full, unclipped image from inside a frame very easy.

So [here’s an unclipped version you can export](9PGX5zb3VKHMYc7qTIv30B-fi-1108%3A38718).

Hope that helps!

Sorry about the lack of Android mocks so far, too 😞

JorgeMucientes committed:
Remove fit end from WC Shipping label banner

JorgeMucientes commented:
Hey @joe-keenan thanks for taking a look a this and updating the new assets. With the new unclipped images I have a question on how should we crop the image to adjust it to the content. 

- `Center crop` Equally crop from top and bottom and keep the image always centered: 

![Screenshot 2022-05-30 at 11 28 29](https://user-images.githubusercontent.com/2663464/170962354-9599a8db-5655-4783-8ad4-f52998cdadee.png)

![Screenshot 2022-05-30 at 11 24 59](https://user-images.githubusercontent.com/2663464/170962019-7db2d1f0-5fc4-499e-8080-27c858e4413c.png)


- `Bottom crop` Crop from bottom of the image (works well only if the text content is longer): 

![Screenshot 2022-05-30 at 11 27 27](https://user-images.githubusercontent.com/2663464/170962209-39eb4db5-1a2e-440f-b60e-4ae55f1e3cb7.png)

![Screenshot 2022-05-30 at 11 26 24](https://user-images.githubusercontent.com/2663464/170962006-180d62cf-fe36-44af-9b90-270bb91ccfb3.png)


Imo `center_cropping` works better specially considering that in most cases the text content is going to be displayed like in the first screenshots of each of the options. 

**EDIT**: Now that I was testing to add the `X` button I realized the `center_cropping` is not going to work well @joe-keenan As the `X` will result overlapping the head of the woman in the image. A possible solution for this is to `bottom_crop` the image and reduce the width of it to a fixed size like 130px for example. This is how it will look like:

![Screenshot 2022-05-30 at 15 09 09](https://user-images.githubusercontent.com/2663464/170999026-c89a4beb-efaf-4541-843d-a6589ea55273.png)

![Screenshot 2022-05-30 at 15 05 11](https://user-images.githubusercontent.com/2663464/170999044-5cb42c2c-5478-424b-93e7-7b4092953b35.png)

I think this works well and the final result is pretty close to the original Figma designs. 

Let me know wyt.

JorgeMucientes committed:
Add unclipped drawables for WC Shipping label banner

JorgeMucientes committed:
Center crop new unclipped images

joe-keenan commented:
Yep, I was initially thinking of centre cropping it, too – but you’re right, with shorter text lengths it will overlap. 
So I’m happy with bottom cropping the way you’ve done here. Nice work!

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailFragment.kt:
```diff
@@ -264,6 +280,10 @@ class OrderDetailFragment : BaseFragment(R.layout.fragment_order_detail), OrderP
     }
 
+    private fun showInstallWcShippingBanner(show: Boolean) {
+        binding.orderDetailInstallWcShippingBanner.isVisible = show && FeatureFlag.WC_SHIPPING_BANNER.isEnabled()
```
❓ What is the reasoning behind adding the feature flag check in the fragment as opposed to doing it in ViewModel?

JorgeMucientes replied:
Good point @anitaa1990 ! The main reason is fixing `unit tests`. From `unit tests` I don't have control if they run on `debug` or `release` buildType. This means that the `OrderDetailViewModelTest` would have different results depending on the `buildType` the tests were run on. A way to overcome this would be to wrap the feature flag inside a class and inject this wrapper class in `OrderViewModelTest` but that seemed overkill. So I ended up moving that to the fragment with the purpose of adding as less code as possible to deal with the feature flag.

anitaa1990 reviewed (approved):
This looks good @JorgeMucientes 🚢  I had one question but it is definitely not blocking. Based on the discussions in the PR, it looks like we're good to merge but I want to confirm so I'll approve it. Feel free to merge it or I can do it once you confirm 👍

JorgeMucientes commented:
Thank you for reviewing all the PR related to the banner @anitaa1990 ! 

Yes this is ready to merge. The latest changes on how to crop the image to support the dismiss `X` button were added in #6664, so this is ready to go!
